mlflow/sklearn/__init__.py

```diff
@@ -884,12 +917,72 @@ def patched_fit(original, self, *args, **kwargs):
                           `sklearn.linear_model.LogisticRegression.fit()` is being patched)
         """
+        global _autolog_training_status
         with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
             if t.should_log():
-                return fit_mlflow(original, self, *args, **kwargs)
+                _autolog_training_status.register_model(self, mlflow.active_run().info.run_id)
+                try:
+                    _autolog_training_status.in_fit_call_scope = True
+                    result = fit_mlflow(original, self, *args, **kwargs)
+                finally:
+                    _autolog_training_status.in_fit_call_scope = False
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_predict(original, self, *args, **kwargs):
+        from mlflow.utils import _inspect_original_var_name
+        global _autolog_training_status
+        with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
+            if t.should_log():
+                result = original(self, *args, **kwargs)
+                eval_dataset = args[0] if len(args) >= 1 else kwargs.get('X')
+                eval_dataset_name = _inspect_original_var_name(eval_dataset)
+                _autolog_training_status.register_predition_result(
+                    self, eval_dataset, result, eval_dataset_name)
+                return result
+            else:
+                return original(self, *args, **kwargs)
+
+    def log_eval_metric(metric_name, metric, metric_api_call_arg_list):
+        if not np.isscalar(metric):
+            # Skip logging metric which is non-scalar value
+            return
+
+        eval_dataset_name, run_id = \
+            _autolog_training_status.get_eval_dataset_name_and_run_id_from_metric_api_arglist(
+                metric_api_call_arg_list
+            )
+        if eval_dataset_name:
+            print('DGB: run patched_metric_api #4')
+            # Note: if the case log the same metric key multiple times,
+            #  newer value will overwrite old value
+            client = mlflow.tracking.MlflowClient()
+
+            # TODO: exception handling ?
+            client.log_metric(
+                run_id=run_id,
+                key=f'{metric_name}_on_{eval_dataset_name}',
+                value=metric
+            )
+
+    def patched_metric_api(original, *args, **kwargs):
+        # TODO:
+        #  check _SklearnTrainingSession and skip logging if in nested runs.
+        global _autolog_training_status
+        metric = original(*args, **kwargs)
+
+        print('DGB: run patched_metric_api #1')
+        if not _autolog_training_status.in_fit_call_scope and \
+                not _autolog_training_status.in_scorer_call_scope and \
+                not _autolog_training_status.in_eval_and_log_metrics_scope:
```

Comment:
Can we just use `_SklearnTrainingSession`?

Reply:
Not a good idea.
When we auto logging metric, we already run out of patched_fit scope, that means we already run out scope of the _SklearnTrainingSession which was created inside patched_fit .

In the latest code , I use _AutologTrainingStatus.dataset_id_to_dataset_name_and_model_id and _AutologTrainingStatus.model_id_to_run_id_map  to track we need autolog the metric into which run_id (or if no related run_id found, then does not autolog metric)